pages/draft-04/migration-notes.md

```diff
@@ -0,0 +1,76 @@
+| `ref` (legacy)    | `ref` (legacy)    | Core          | Applicator        | -                                                                                                                                                                      |
+| Not present       | `allOf`           | Core          | Applicator        | -                                                                                                                                                                      |
+| Not present       | `anyOff`          | Core          | Applicator        | -                                                                                                                                                                      |
+| Not present       | `definitions`     | Validation    | Reserved Location | -                                                                                                                                                                      |
+| Not present       | `maxProperties`   | Validation    | Assertion         | -                                                                                                                                                                      |
+| Not present       | `minProperties`   | Validation    | Assertion         | -                                                                                                                                                                      |
+| Not present       | `not`             | Core          | Applicator        | -                                                                                                                                                                      |
+| Not present       | `oneOf`           | Core          | Applicator        | -                                                                                                                                                                      |
+
+<Infobox label="Note"> Starting with Draft 4, schema identifiers that use an empty URI "" or a fragment-only URI "#" are no longer allowed.
+
+In Draft 3, these identifiers were considered valid:
+
+```
+id: ""
+id: "#"
+```
+
+However, this format is now prohibited from Draft 4 onwards.
+</Infobox>
+
+<Infobox label="Note">
+Before Draft 3, the JSON Schema only included the Core specification, which outlined the foundational elements for schema structure. With Draft 4, the specification expanded to incorporate Validation, establishing rules for data format, structure, and type requirements. This addition enabled schema authors to define and enforce validation constraints directly within their schemas, making it easier for implementers to ensure data integrity and compatibility as schemas evolved.
```

Comment:
>  Everything that we now consider Core, Validation, and Hyper-Schema was included in that one spec

What was the one spec?

If you also stated that there was originally only one specification, at what point did the Validation become necessary?

> In draft-04, the one spec was split into the three parts we're familiar with today. So, the spec didn't expand in draft-04 to include validation. Validation was always there. It just split that content into a different document.

Thank you. I will phrase this better.

> The Content vocabulary and a few Meta-Data keywords moved from Hyper-Schema to Validation (draft-07) and the Applicator vocabulary moved from Validation to Core (2019-09).

Are there any more data for scenarios like this? 

To better represent this Spec note, what would be the most nearly accurate detail?

Reply:
> What was the one spec?

I'm not sure what you mean. Draft-01, draft-02, and draft-03 were all just one spec that included what is now Core, Validation, and Hyper-Schema. It didn't really have another name like the three do now. It was just *the* JSON Schema spec.

> If you also stated that there was originally only one specification, at what point did the Validation become necessary?

Validation was always part of the spec. It was included from the first draft. It was always necessary.

> Are there any more data for scenarios like this?

I think this kind of information is in the release notes for each draft. The older drafts don't have release notes, but those didn't have these kinds of changes.

> To better represent this Spec note, what would be the most nearly accurate detail?

Honestly, I'm not sure what this spec note is trying to communicate. Is it just that there are now three specs instead of one?[^1] That was just a change in the way the spec was maintained. It has little affect on what people can and can't do with JSON Schema. However, splitting off Hyper-Schema meant draft-04 was the first draft where you could implement JSON Schema without including Hyper-Schema support. Implementers could implement Core and Validation and treat Hyper-Schema as an extension that they don't support.

[^1]: Actually, it was broken into four specs: Core, Validation, Hyper-Schema, and JSON Reference. But, JSON Reference wasn't maintained by us. JSON Reference was merged back into Core in draft-06.

Reply:
Thank you for clarifying about the initial "JSON Schema Spec".